docs/design/per-op-config.md

```diff
@@ -0,0 +1,179 @@
+# per-op config
+* **Type**: Design
+* **Author(s)**: AWS SDK for Kotlin team
+
+# abstract
+The smithy-kotlin project generates clients with the ability to perform operation- or block-scoped config overrides.
+This functionality is exposed by supporting a `withConfig` method on clients, which takes a config builder and returns a
+new "subclient" which performs operations using the modified config. We also outline several proposed design
+alternatives that were rejected.
+
+# design
+All code samples in this document assume the following basic client interface. The client has an arbitrary number of operations,
+the declarations of which are omitted.
+This particular client has two config
+fields: a logging mode and an HTTP client to be used to make requests. Note that HTTP clients (`HttpClientEngine`) in
+the smithy-kotlin runtime implement `Closeable`.
+
+```kotlin
+interface ServiceClient : SdkClient {
+    companion object {
+        fun invoke(block: Config.Builder.() -> Unit): ServiceClient { /* returns instance of implementation class DefaultServiceClient */ }
+    }
+    
+    class Config private constructor(builder: Builder) {
+        val logMode: SdkLogMode = builder.logMode ?: SdkLogMode.Default
+        val httpEngine: HttpClientEngine = builder.httpEngine
+        
+        class Builder {
+            var logMode: SdkLogMode? = null
+            val httpEngine: HttpClientEngine? = null
+            fun build(): Config = Config(this)
+        }
+    }
+    
+    // suspend fun operationA, operationB, etc...
+}
+
+class DefaultServiceClient(config: ServiceClient.Config) : ServiceClient { /* ... */ }
+```
+
+Client interfaces receive a new `withConfig` extension, which takes a config builder and returns a "child" client with
+any fields specified in the builder applied to its config.
+
+```kotlin
+fun ServiceClient.withConfig(block: ServiceClient.Config.Builder.() -> Unit): ServiceClient { /* ... */ }
+```
+
+In practice, to perform one or more operations with a modified config in-place, it is most effective to make your call
+within a `use()` block on the returned client to have the `close()` handled for you. It's still possible to pass the
```

Comment:
Agreed. Instead of in " in practice" I would just spell out the semantics here.

e.g. something like

> The client returned from `withConfig` is independent from the original client. It has it's own lifetime and should be closed when no longer needed. Any closeable resources will be kept alive until both clients are finished with them.